source/extensions/health_checkers/thrift/thrift.cc

```diff
@@ -0,0 +1,119 @@
+      protocol_(ProtoUtils::getProtocolType(thrift_config.protocol())),
+      client_factory_(client_factory) {
+  if (transport_ == TransportType::Auto || protocol_ == ProtocolType::Auto ||
+      protocol_ == ProtocolType::Twitter) {
+    throw EnvoyException(
+        fmt::format("Invalid thrift health check configuration: {}", thrift_config.DebugString()));
+  }
+}
+
+ThriftHealthChecker::ThriftActiveHealthCheckSession::ThriftActiveHealthCheckSession(
+    ThriftHealthChecker& parent, const Upstream::HostSharedPtr& host)
+    : ActiveHealthCheckSession(parent, host), parent_(parent),
+      hostname_(getHostname(host, parent_.cluster_.info())) {
+  ENVOY_LOG(trace, "ThriftActiveHealthCheckSession construct hostname={}", hostname_);
+}
+
+ThriftHealthChecker::ThriftActiveHealthCheckSession::~ThriftActiveHealthCheckSession() {
+  ENVOY_LOG(trace, "ThriftActiveHealthCheckSession destruct");
+  ASSERT(client_ == nullptr);
+}
+
+void ThriftHealthChecker::ThriftActiveHealthCheckSession::onDeferredDelete() {
+  if (client_) {
+    expect_close_ = true;
+    client_->close();
+  }
+}
+
+void ThriftHealthChecker::ThriftActiveHealthCheckSession::onInterval() {
+  ENVOY_LOG(trace, "ThriftActiveHealthCheckSession onInterval");
+  if (!client_) {
+    ENVOY_LOG(trace, "ThriftActiveHealthCheckSession construct client");
+    client_ = parent_.client_factory_.create(
+        *this, parent_.transport_, parent_.protocol_, parent_.method_name_, host_,
+        /* health checker seq id */ 0, /* fixed_seq_id */ true);
+    client_->start();
+    expect_close_ = false;
+  }
+
+  client_->sendRequest();
+}
+
+void ThriftHealthChecker::ThriftActiveHealthCheckSession::onTimeout() {
+  if (client_) {
+    expect_close_ = true;
+    client_->close();
+  }
+}
+
+void ThriftHealthChecker::ThriftActiveHealthCheckSession::onResponseResult(bool is_success) {
+  if (is_success) {
+    handleSuccess();
+  } else {
+    // TODO(kuochunghsu): We might want to define retriable response.
+    handleFailure(envoy::data::core::v3::ACTIVE, /* retriable */ false);
+  }
+
+  if (client_ && !parent_.reuse_connection_) {
+    expect_close_ = true;
+    client_->close();
+  }
+}
+
+Upstream::Host::CreateConnectionData
+ThriftHealthChecker::ThriftActiveHealthCheckSession::createConnection() {
+  return host_->createHealthCheckConnection(parent_.dispatcher_, parent_.transportSocketOptions(),
+                                            parent_.transportSocketMatchMetadata().get());
+}
+
+void ThriftHealthChecker::ThriftActiveHealthCheckSession::onEvent(Network::ConnectionEvent event) {
+  if (event == Network::ConnectionEvent::RemoteClose ||
+      event == Network::ConnectionEvent::LocalClose) {
+    ENVOY_LOG(trace, "on event close, is_local_close={} expect_close={}",
+              event == Network::ConnectionEvent::LocalClose, expect_close_);
+    if (!expect_close_) {
+      handleFailure(envoy::data::core::v3::NETWORK);
```

Comment:
Can this happen some time after a full response if the server has some kind of idle timeout? And if so, is that really a failure? The HTTP health checker might have a useful pattern to follow here.

Reply:
Do you mean race condition like `onTimeout` called first and a full response is on the fly?
If so, onTimeout will close the connection and no longer to receive connection callback.
Other health checkers use the same trick: a `expect_*` flag to prevent race condition.